ethereum_history_api/circuits/lib/src/rlp/rlp_type.nr

```diff
@@ -0,0 +1,49 @@
+use crate::misc::fragment::Fragment;
+use crate::rlp::decode::{decode_to_rlp_header, copy_subarray, RlpHeader, LIST, STRING};
+use crate::rlp::fragment::RlpFragment;
+
+struct FragmentWithDataType<MAX_BYTES_LEN> {
+    fragment: Fragment<MAX_BYTES_LEN>,
+    data_type: u64 // STRING or LIST
+}
+
+type RlpList<MAX_BYTES_LEN, MAX_FIELDS> = BoundedVec<FragmentWithDataType<MAX_BYTES_LEN>, MAX_FIELDS>;
+
+/// Returns an RLP list look-up table.
+/// For string elements, the offset points to the payload, whereas the offset
+///   of a list element points to the RLP header of that element.
+pub fn decode_list<MAX_BYTES_LEN, MAX_FIELDS>(input: [u8; MAX_BYTES_LEN]) -> RlpList<MAX_BYTES_LEN, MAX_FIELDS> {
+    let mut num_fields: u64 = 0;
+    let mut rlp_list: RlpList<MAX_BYTES_LEN, MAX_FIELDS> = BoundedVec::new();
+
+    let RlpHeader {offset, length, data_type} = decode_to_rlp_header(input);
+    assert(data_type == LIST, "Expected a list");
+
+    let total_len = length + offset;
+    assert(total_len <= input.len(), "Total length exceeds input length");
+
+    let mut current_offset = offset;
+    for _ in 0..MAX_FIELDS {
+        if (current_offset < total_len) {
+            let header: [u8; MAX_BYTES_LEN] = copy_subarray(input, current_offset);
+
+            let RlpHeader {offset: field_off, length: field_len, data_type: field_type} = decode_to_rlp_header(header);
+
+            let mut fragment: FragmentWithDataType<MAX_BYTES_LEN> = FragmentWithDataType {
+                fragment: Fragment {
+                    offset: current_offset + (1 - field_type) * field_off, // If the ith slot contains a list, include its RLP header.
+                    length: field_len + field_type * field_off, // If the ith slot contains a list, include the length of its header.
+                    data: input
+                },
+                data_type: field_type
```

Comment:
I think we can reuse the data_type name here as the variable above is already used and asserted. Seems like introducing another name raises more questions than we avoid by not shadowing

Reply:
Changed